In [ ]:
import numpy as np
import pandas as pd
import plotly
import plotly.figure_factory as ff
import plotly.graph_objs as go
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import SGDClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
init_notebook_mode(connected=True)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Assignment_8/task_b.csv')
data=data.iloc[:,1:]

In [ ]:
data.head()

f1            f2        f3    y
0  -195.871045 -14843.084171  5.532140  1.0
1 -1217.183964  -4068.124621  4.416082  1.0
2     9.138451   4413.412028  0.425317  0.0
3   363.824242  15474.760647  1.094119  0.0
4  -768.812047  -7963.932192  1.870536  0.0

In [ ]:
data.corr()['y']

f1    0.067172
f2   -0.017944
f3    0.839060
y     1.000000
Name: y, dtype: float64

In [ ]:
data.std()

f1      488.195035
f2    10403.417325
f3        2.926662
y         0.501255
dtype: float64

In [ ]:
X=data[['f1','f2','f3']].values
Y=data['y'].values
print(X.shape)
print(Y.shape)

(200, 3)
(200,)


#Task1.1


In [ ]:
log_clf = SGDClassifier(loss='log',shuffle = False)#shuffle is made false to avoid random_state
log_clf.fit(X,Y)
log_clf.coef_

array([[  383.17082929,   -67.9141977 , 14465.32903487]])

#Task1.2

In [ ]:
SVM_clf = log_clf.set_params(loss='hinge')
SVM_clf.fit(X,Y)
SVM_clf.coef_

array([[  432.63085834,   -81.36713692, 14466.94777939]])

##Observations: Before standardization
1. In logistic regression the feature impotance is 
        F3>F1>F2
2. In SVM the feature importance is 
        F3>F1>F2
3. F3 has highest correlation with the class label 'Y' so it got more feature importance. That means changes in F3 highly cause changes in "Y".

#Task2.1

In [ ]:
sclr = StandardScaler()
standard_X= sclr.fit_transform(X)

log_clf.fit(standard_X,Y)
log_clf.coef_

array([[-3.11077234, -0.3901838 ,  9.46374622]])

In [ ]:
SVM_clf.fit(standard_X,Y)
SVM_clf.coef_

array([[-3.11077234, -0.3901838 ,  9.46374622]])

##Observations: After standardization
1. Both models return same wieght vector indicating same feature importance
        F3>F1>F2
2. There is no change in feature importance before and after standardization.
3. Logistic regression is not effected by variance of the features.
4. Since SVM is a distance based classifier, it must be effected by variance of features, but here no 
effect is observed.
5. We can see that correaltion with class label "Y" is of the same order
        F3(0.83)>F1(0.067)>F2(-0.01)
6. So feature importance is highly effected by class correlation.
        

#Ref:
1. https://builtin.com/data-science/when-and-why-standardize-your-data
2. https://stackoverflow.com/questions/52905164/class-correlation-and-its-effects